rectools/models/bert4rec.py

```diff
@@ -0,0 +1,197 @@
+import typing as tp
+from typing import List, Tuple
+
+import numpy as np
+import torch
+from pytorch_lightning import Trainer
+from torch import nn
+
+from rectools.models.sasrec import (
+    CatFeaturesItemNet,
+    IdEmbeddingsItemNet,
+    ItemNetBase,
+    LearnableInversePositionalEncoding,
+    PointWiseFeedForward,
+    PositionalEncodingBase,
+    SessionEncoderDataPreparatorBase,
+    SessionEncoderLightningModule,
+    SessionEncoderLightningModuleBase,
+    TransformerLayersBase,
+    TransformerModelBase,
+)
+
+PADDING_VALUE = "PAD"
+MASKING_VALUE = "MASK"
+
+
+class BERT4RecDataPreparator(SessionEncoderDataPreparatorBase):
+    """TODO"""
+
+    def __init__(
+        self,
+        session_max_len: int,
+        batch_size: int,
+        dataloader_num_workers: int,
+        train_min_user_interactions: int,
+        mask_prob: float,
+        item_extra_tokens: tp.Sequence[tp.Hashable],
+        shuffle_train: bool = True,
+    ) -> None:
+        super().__init__(
+            session_max_len=session_max_len,
+            batch_size=batch_size,
+            dataloader_num_workers=dataloader_num_workers,
+            train_min_user_interactions=train_min_user_interactions,
+            item_extra_tokens=item_extra_tokens,
+            shuffle_train=shuffle_train,
+        )
+        self.mask_prob = mask_prob
+
+    def _mask_session(self, ses: List[int]) -> Tuple[List[int], List[int]]:
+        masked_session = ses.copy()
+        target = ses.copy()
+        random_probs = np.random.rand(len(ses))
+        for j in range(len(ses)):
+            if random_probs[j] < self.mask_prob:
+                random_probs[j] /= self.mask_prob
+                if random_probs[j] < 0.8:
+                    masked_session[j] = 1
```

Comment:
Let's make it more clear and safe.
In the beginning:
mask_id = self.item_id_map.convert_to_internal([MASKING_VALUE])[0]
Here:
masked_session[j] = mask_id

Reply:
Training is significantly slower

Reply:
fixed